components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java

```diff
@@ -0,0 +1,213 @@
+    public ServiceURLBuilder addPath(String ... paths) {
+
+        this.urlPaths = paths;
+        return this;
+    }
+
+    /**
+     * This method is called when the URL needs to be appended with other parameters. Such parameters should be
+     * entered as <k,v> pairs.
+     *
+     * @param key   Key.
+     * @param value Value.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder addParameter(String key, String value) {
+
+        parameters.put(key, value);
+        return this;
+    }
+
+    /**
+     * This method is called when the URL needs to be appended with a fragment.
+     *
+     * @param fragment Fragment.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder setFragment(String fragment) {
+
+        this.fragment = fragment;
+        return this;
+    }
+
+    /**
+     * This method is called when the URL fragment needs to be appended with parameters. Such parameters should be
+     * entered as <k,v> pairs. These parameters will get appended with an "&".
+     *
+     * @param key   Key.
+     * @param value Value.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder addFragmentParameter(String key, String value) {
+
+        fragmentParams.put(key, value);
+        return this;
+    }
+
+    @Override
+    public ServiceURL build() throws URLBuilderException {
+
+        String protocol = fetchProtocol();
+        String hostName = fetchHostName();
+        int port = fetchPort();
+        String resolvedUrlContext = buildUrlPath(urlPaths);
+        String resolvedFragment = buildFragment(fragment, fragmentParams);
+        String tenantDomain = resolveTenantDomain();
+
+        StringBuilder resolvedUrlStringBuilder = new StringBuilder();
+        if (isTenantQualifiedUrlsEnabled()) {
+            if (StringUtils.isNotBlank(tenantDomain)) {
+                resolvedUrlStringBuilder.append("/t/").append(tenantDomain);
+            }
+        }
+
+        if (StringUtils.isNotBlank(resolvedUrlContext)) {
+            if (resolvedUrlContext.trim().charAt(0) != '/') {
+                resolvedUrlStringBuilder.append("/").append(resolvedUrlContext.trim());
+            } else {
+                resolvedUrlStringBuilder.append(resolvedUrlContext.trim());
+            }
+        }
+
+        return new ServiceURL(protocol, hostName, port, resolvedUrlStringBuilder.toString(), parameters,
+                resolvedFragment);
+    }
+
+    private String resolveTenantDomain() {
+
+        String tenantDomain = IdentityTenantUtil.getTenantDomainFromContext();
+        if (StringUtils.isBlank(tenantDomain)) {
+            tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        }
+        return tenantDomain;
+    }
+
+    private String buildFragment(String fragment, Map<String, String> fragmentParams) {
+
+        if (StringUtils.isNotBlank(fragment)) {
```

Comment:
Can there be a situation where we will have both the fragment string and fragment params?

Reply:
The spec for URIs [1] says the semantics of the fragment are considered unknown and are effectively unconstrained. So, if we want to accommodate such a case, we can have it. But, we need to rethink the fragment structure then.. 

ref: []()https://tools.ietf.org/html/rfc3986#section-3.5

Reply:
Let's not have both. setFragment() denotes it overrides. Common practice followed